File: 백준/Gold/1068. 트리/트리.cc
#include <bits/stdc++.h>
using namespace std;

vector<int> v[100001];
bool visited[100001];

int n, root, del, cnt=0;

void delVisted(int index) {
	visited[index]=true;	
	
	for(int i=0; i<v[index].size(); i++) {
		int next = v[index][i];
		if(!visited[next]) {
			delVisted(next);
		}
	}
}

void answer(int index) {
	visited[index]=true;
	
	int child=0;
	for(int i=0; i<v[index].size(); i++) {
		int next = v[index][i];
		if(!visited[next]) {
			answer(next);
			child++;
		}
	}
	if(child>0) return;
	else cnt++;
}

int main() {
	scanf("%d", &n);
	int x;
	for(int i=0; i<n; i++) {
		scanf("%d", &x);
		if(x==-1) root=i;
		else {
			v[x].push_back(i);
		}
	}	 
	
	scanf("%d", &del);
	delVisted(del);
	
	if(root==del) cout << 0;
	else {
		answer(root);
		cout << cnt;
	}

	return 0; 
}

File: 백준/Gold/1068. 트리/트리.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());

        st = new StringTokenizer(br.readLine());
        Map<Integer, List<Integer>> tree = new HashMap<>();

        for (int i = 0; i < n; i++) {
            tree.put(i, new ArrayList<>());
        }

        int root = -1;
        for (int i = 0; i < n; i++) {
            int x = Integer.parseInt(st.nextToken());
            if (x == -1) {
               root = i;
            } else {
                tree.get(x).add(i);
            }
        }

        boolean[] visited = new boolean[n];
        int deleteNode = Integer.parseInt(br.readLine());
        int answer = DFS(root, deleteNode, visited, tree);
        System.out.println(answer);
    }


    private static int DFS(int node, int deleteNode, boolean[] visited, Map<Integer, List<Integer>> tree) {
        visited[node] = true;
        if (node == deleteNode) return 0; // 루트가 삭제 된 경우

        int result = 0;
        int childCount = 0;
        for (int child : tree.get(node)) {
            if (child == deleteNode) continue;

            if (!visited[child]) {
                childCount++;
                result += DFS(child, deleteNode, visited, tree);
            }
        }

        if (childCount == 0) return result + 1;
        return result;
    }

}
